src/lambda_toolshed/papillon.cljc

```diff
@@ -102,19 +95,19 @@
   [ctx]
   (if (satisfies? ReadPort ctx)
     (go (leave (<! ctx)))
-    (let [stack (::stack ctx)]
-      (if-let [ix (peek stack)]
-        (recur (-> ctx
-                   (update ::stack pop)
-                   (try-stage ix (if (::error ctx) :error :leave))))
-        ctx))))
+    (if-let [ix (peek (::stack ctx))]
+      (recur (-> ctx
+                 (update ::stack pop)
+                 (try-stage ix (if (::error ctx) :error :leave))))
+      ctx)))
 
 (defn- init-ctx
   "Inialize the given context `ctx` with the necessary data structures to
   process the interceptor chain `ixs`."
   [ctx ixs]
-  (assoc (enqueue ctx ixs)
-         ::stack []))
+  (-> (assoc (enqueue ctx ixs)
+             ::stack [])
+      (vary-meta assoc :type ::ctx)))
```

Comment:
for further consistency with the rest of the code (eg. `enter` & `leave`), it would worth using the `(-> ctx ...)` structure:
```clojure
(defn- init-ctx
  "Inialize the given context `ctx` with the necessary data structures to
  process the interceptor chain `ixs`."
  [ctx ixs]
  (-> ctx
      (enqueue ixs)
      (assoc ::stack [])
      (vary-meta assoc :type ::ctx)))
```